libui/src/ui.rs

```diff
@@ -246,3 +266,183 @@ impl<'s> EventLoop<'s> {
         }
     }
 }
+
+/// The struct to enqueue a closure from another thread
+pub struct EventQueue {
+    initialized: Arc<RwLock<bool>>,
+}
+
+impl EventQueue {
+    /// Tries to enqueue the callback to event queue for the main thread
+    ///
+    /// Returns false if the `UI` is already dropped and unable to queue the event.
+    pub fn queue_main<F: FnOnce() + 'static + Send>(&self, callback: F) -> bool {
+        let initialized = self.initialized.read().unwrap();
+
+        if !*initialized {
+            return false;
+        }
+
+        extern "C" fn c_callback<G: FnOnce()>(data: *mut c_void) {
+            unsafe {
+                from_void_ptr::<Option<G>>(data).take().map(|f| f());
+            }
+        }
+
+        unsafe {
+            libui_ffi::uiQueueMain(Some(c_callback::<F>), to_heap_ptr(Some(callback)));
+        }
+
+        true
+    }
+}
+
+/// The struct to enqueue a closure from another thread,
+/// with holding data that is `!Send`.
+/// 
+/// **Example**
+/// 
+/// ```
+/// use std::cell::RefCell;
+/// use std::rc::Rc;
+/// use libui::controls::Label;
+/// # use libui::{EventQueueWithData, UI};
+/// # fn heavy_process() -> String {
+/// #     "Success".into()
+/// # }
+/// # 
+/// # fn example(ui: &UI) {
+///
+/// // label is !Send
+/// let label = Rc::new(RefCell::new(Label::new("processing...")));
+/// let queue_with_data = EventQueueWithData::new(ui, label);
+///
+/// std::thread::spawn(move || {
+///     let result = heavy_process();
+///     queue_with_data.queue_main(move |label| {
+///         // we can access label inside queue_main on main thread!
+///         label.borrow_mut().set_text(&result);
+///     })
+/// });
+///
+/// # }
+/// ```
```

Comment:
```suggestion
/// ```
///
/// **Note**
/// The data held by this struct will be leaked if 
/// libui is uninitialized when this struct is dropped.
```

It might be better to note there is possibility for leak data